File: packages/schemas/python/simulation.py
from __future__ import annotations
from typing import Literal, Optional
from pydantic import BaseModel


class FlightState(BaseModel):
    flight_id: str
    status: Literal["scheduled", "delayed", "cancelled", "diverted", "airborne"]
    delay_minutes: int = 0
    cascade_order: int = -1         # -1=unaffected, 0=direct, 1,2=cascade
    p_delayed: float = 0.0          # probability 0–1
    reason: Optional[str] = None
    tail: Optional[str] = None
    origin: Optional[str] = None
    destination: Optional[str] = None
    new_departure: Optional[str] = None


class CascadeSummary(BaseModel):
    directly_affected: int = 0
    cascade_1: int = 0
    cascade_2: int = 0
    total_affected: int = 0
    total_delay_minutes: Optional[int] = None
    cancellation_cost_usd: Optional[float] = None


class SimulationState(BaseModel):
    sim_time: Optional[str] = None
    active_events: list = []
    flight_states: dict[str, FlightState] = {}
    recovery_plans: list = []
    schedule: list = []

File: packages/schemas/src/simulation.ts
// ── Simulation state ──────────────────────────────────────────────────────────

export interface FlightState {
  flight_id: string
  status: "scheduled" | "delayed" | "cancelled" | "diverted" | "airborne"
  delay_minutes: number
  cascade_order: number         // -1=unaffected, 0=direct, 1=cascade-1, 2=cascade-2
  p_delayed: number             // probability 0–1
  reason?: string
  tail?: string
  origin?: string
  destination?: string
  new_departure?: string        // ISO-8601 if rescheduled
}

export interface CascadeSummary {
  directly_affected: number
  cascade_1: number
  cascade_2: number
  total_affected: number
  total_delay_minutes?: number
  cancellation_cost_usd?: number
}

export interface SimulationState {
  sim_time?: string            // ISO-8601
  active_events: import("./events").ActiveEvent[]
  flight_states: Record<string, FlightState>
  recovery_plans: import("./recovery").RecoveryPlan[]
  schedule: import("./network").ScheduledFlight[]
}

// ── WebSocket message envelopes ───────────────────────────────────────────────

export type WSMessageType =
  | "ping"
  | "pong"
  | "state_snapshot"
  | "flight_update"
  | "event_triggered"
  | "recovery_plans"
  | "cascade_update"

export interface WSMessage<T = unknown> {
  type: WSMessageType
  payload?: T
  timestamp?: string
}

export interface FlightUpdatePayload {
  flight_states: Record<string, FlightState>
  cascade_summary?: CascadeSummary
  event?: import("./events").ActiveEvent
  active_events?: import("./events").ActiveEvent[]
  recovery_plans?: import("./recovery").RecoveryPlan[]
  schedule?: import("./network").ScheduledFlight[]
}
